### ISSUE TITLE
arXiv '21 | ZeRO-Infinity: Breaking the GPU Memory Wall for Extreme Scale Deep Learning

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/read-papers#196]
https://arxiv.org/pdf/2104.07857.pdf

周末看.

**Comment by vycezhong** [comment]
这文章写得不错. 

## Background & Motivation 

模型增长速度远远超过了GPU显存的增长速度. GPU memory wall(内存墙)成为了训练大模型的阻碍. 

目前训练Transformer-based大模型的de facto做法: 
- 3D parallelism (data, model, pipeline)
- mixed-precision training. (用tensor core加速训练, 但对于如此大的模型其实增加了显存) 
- activation checkpoint (recomputation)
- gradient accumulation (额外好处: 减少通信. 坏处: 需要FP32 copy of gradients)
- Adam optimizer. 

其他一些做法:
- ZeRO optimizer (remove memory redundancies in data parallelism).  包括optimizer states, gradients和parameters. 
- ZeRO offload. 其实和PS没啥区别...就是把FP32的parameters, optimizer states放在CPU上, 在CPU上做更新. GPU -> CPU: FP16 gradients. CPU -> GPU: FP16 parameters.

## Memory Requirement

1. model states: 对于mixed-precision training, 每个参数需要20 bytes. (FP16 parameters, FP16 gradients, FP32 parameters, FP32 momentum, FP32 variance, FP32 gradients). 最后一个FP32 gradients是为了gradient accumulation. 

而一个transformer block的参数量由两个超参数决定: hidden dimension (hd), 和layers (hl). 一个决定宽度, 一个决定深度. 注意number of attention heads其实是不决定参数数量的. 

一个transformer block的参数主要有四个部分:
- attention projection matrix: (hd, 3hd)
- attention linear layer: (hd, hd)
- feedforward layer: (hd, 4hd) + (4hd, hd)

所以参数量一共 12 x nl x hd^2 

所以总大小为 240 x nl x hd^2 (bytes)

2. residual states. 主要是activation memory. activation checkpoint一般保存transformer block的输出, 所以总的大小为

2 x bsz x seq x hd x nl / ci

ci是两个activation checkpoint间的transformer block个数(比如如果保存每个transformer block的activation, ci=1).  由于是FP16的activation, 所以要乘以2. 


3. Model State Working Memory (MSWM). 最大的operator所需要的显存大小. 比如在transformer-based model里面, 最大的operator是feedforward layer, 需要4 x hd x 4hd bytes的连续内存. 常见解决方法是model parallelism. 

4. Activation Working Memory (AWM). 两个activation checkpoint间所需要的activation大小. 

2 x bsz x seq x ci x (8 x hd + attn_heads x seq)

括号外面的比较好理解. 括号里面的attn_heads x seq这一项也好理解, 但 8 x hd我真的理解不了. 我觉得不止这么多... (我觉得有3 + 1 + 1 + 1 + 4 + 1 = 11)

**Comment by vycezhong** [comment]
## Bandwidth Requirements

提出了一个AIT的概念， 以前还没有听说过. 

ait = total_computation / total_data_movement 

然后efficiency = compute_time / (compute_time + communication_time)

compute_time = total_computation / peak_{tp}. peak_[tp}是peak computational throughput.

communication_time = total_data_movement / bw = total_computation / ait / bw

所以efficiency = ait * bw / (ait * bw + peak_{tp}). 

显然, bw和peak_{tp}都是已知的. 所以只需要量化ait.

**Comment by vycezhong** [comment]
## System Design

我对其细节非常失望. 基本没有讲清楚. 

![image](https://user-images.githubusercontent.com/25879526/121111927-4a026600-c842-11eb-97c1-8ea04c20fde7.png)
先看DGX-2 cluster的硬件情况. 其中感觉CPU memory这个似乎不太对...怎么会只有3GB/s... 第一行的12GB/s差不多. 就算用memcpy做benchmark最差也有5GB/s吧. 如果加上omp可以达到几十GB/s. 

跨机的通信能做到60GB/s??? 应该是多个100Gbps的网卡一起算了吧? 

1. offload engine for model states. 按需offload / prefetch. 

2. offload for activations. 10T参数的模型的activations有0.76TB... 说实话也不小，但足够存在CPU memory中. 

3. memory-centric filing for working memory. 就是一个很大的operator, 原本我们是用model parallelism来做的. 但其实可以变成一个sequential的操作. 确实. 算好一个tile就release对应的参数和输入. 

但做了这么多offload是有问题的. 因为CPU memory bandwidth是比GPU memory bandwidth低一个数量级的. 而disk的bandwidth又比CPU memory bandwidth低一个数量级. 

问题的关键是PCIe带宽太低了 (12GB/s).  如果要做CPU offload，就需要足够大的batch size来提高AIT. 另外还可以与GPU计算做overlap. 

实现上用了很多pytorch hooks. 

后面还有一些tricks. 比如bandwidth-centric partitioning. 我觉得很trivial...就是把原来的一个node存所有参数改成所有node都存一部分参数(我以为之前就是这样的).  这样传输开销就分摊了. 但看看文章是怎么吹这个点的:

> bandwidth-centric partitioning: a novel data mapping and parallel data retrieval strategy for offloaded parameters and gradients that allows ZeRO-Infinity to achieve virtually unlimited heterogeneous memory bandwidth (details in Sec. 6.1),

emmmmmm....

还有一个overlap centric design. emmm. 看上去就是早点做prefetch. 

> For instance, before executing the 𝑖 𝑡ℎ operator, the prefetcher can invoke nc, cg, and gg-transfer for parameters required by 𝑖 + 3, 𝑖 + 2, and 𝑖 + 1 operators, respectively.

最后说了automating data movement. 就是在执行op前，如果发现参数没到位，就allgather. 执行op后，做partition和可选的offload. 具体得我去用了deepspeed才知道是不是真自动化. 



## Evaluation 

这实验是什么鬼. super-linear???? 文章的解释是PCIe和NVMe bandwidth越来越多了. offloading的开销减小了. 额.

![image](https://user-images.githubusercontent.com/25879526/121124025-cf901100-c856-11eb-99e5-8a675d450845.png)


注意一下这个图，做offloading的开销. 当hidden size越大，offload所带来的开销（纵轴的意思是降低throughput by xx times）越来越小
![image](https://user-images.githubusercontent.com/25879526/121124382-678dfa80-c857-11eb-8f32-be21c7796bb6.png)

**Comment by vycezhong** [comment]
文章看完了，好像没提到底是怎么做并行的？好像是没用到model parallelism. pipeline细节也没提(难不成没用到?). 所以完全靠data parallelism + ZeRO的各种partition? 啊这.
